### PR TITLE
filter: Add --empty-results-reporting={error,warn,skip} option

### PR DESCRIPTION
**Pull request by joverlee521**
### Description of proposed changes
Allows users to specify reporting behavior if filtering produces an empty result. The default behavior is still to raise an error if there's an empty result.

### Related issue(s)

Related to https://github.com/nextstrain/seasonal-flu/pull/76 (see [discussion on Slack](https://bedfordlab.slack.com/archives/C03KWDET9/p1677875939132259?thread_ts=1677792435.138489&cid=C03KWDET9))

### Testing
Added Cram test for new `--empty-results-reporting` option. 

### Checklist

- [x] Add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) summarizing the changes in this PR that are end user focused. Keep headers and formatting consistent with the rest of the file.

### CODE REVIEW

File: augur/filter/__init__.py [hunk 1]
```diff
@@ -68,6 +69,12 @@ def register_arguments(parser):
     output_group.add_argument('--output-metadata', help="metadata for strains that passed filters")
     output_group.add_argument('--output-strains', help="list of strains that passed filters (no header)")
     output_group.add_argument('--output-log', help="tab-delimited file with one row for each filtered strain and the reason it was filtered. Keyword arguments used for a given filter are reported in JSON format in a `kwargs` column.")
+    output_group.add_argument(
+        '--empty-results-reporting',
```
Comment (victorlin) [inline augur/filter/__init__.py:73]: This option doesn't fit under the "outputs" group:

> outputs:
>   possible representations of filtered data (at least one required)

Maybe it would fit better under the default "options" group?

```suggestion
    parser.add_argument(
        '--empty-results-reporting',
```

File: augur/filter/__init__.py [hunk 2]
```diff
@@ -68,6 +69,12 @@ def register_arguments(parser):
...
+        type=EmptyResultsReportingMethod,
+        choices=[ method for method in EmptyResultsReportingMethod ],
```
Comment (tsibley) [inline augur/filter/__init__.py:75]: _non-blocking_

Does

```suggestion
        choices=list(EmptyResultsReportingMethod),
```

work?

File: augur/filter/__init__.py (hunk 1 above)
Comment (joverlee521) [reply augur/filter/__init__.py:73]: > Maybe it would fit better under the default "options" group?

Hmm, I would look for options related to the outputs in the `outputs` group. 
I can reword the argument group description to make it clearer?

Comment (joverlee521) [reply augur/filter/__init__.py:73]: > Though maybe the option name should use "output" instead of "results", e.g. --empty-output-reporting?

I specifically named the option with "results" since we have the `--output-log` option, which is not taken into account by the `--empty-results-reporting` option. That might just be me being nitpicky about words 🤷

File: augur/filter/__init__.py (hunk 2 above)
Comment (joverlee521) [reply augur/filter/__init__.py:75]: Ah, yup it does work. The list comprehension was left over from when the class was just a basic `Enum` instead of the custom `ArgparseEnum`.

Comment (joverlee521) [reply augur/filter/__init__.py]: Updated in force push.

File: augur/filter/__init__.py (hunk 1 above)
Comment (joverlee521) [reply augur/filter/__init__.py:73]: Updated description in force push:
```
outputs:
  options related to outputs, at least one of the possible representations
  of filtered data (--output, --output-metadata, --output-strains) is
  required

  --output OUTPUT, --output-sequences OUTPUT, -o OUTPUT
                        filtered sequences in FASTA format (default: None)
  --output-metadata OUTPUT_METADATA
                        metadata for strains that passed filters (default:
                        None)
  --output-strains OUTPUT_STRAINS
                        list of strains that passed filters (no header)
                        (default: None)
  --output-log OUTPUT_LOG
                        tab-delimited file with one row for each filtered
                        strain and the reason it was filtered. Keyword
                        arguments used for a given filter are reported in JSON
                        format in a `kwargs` column. (default: None)
  --empty-results-reporting {error,warn,skip}
                        How should empty filter results be reported. (default:
                        error)
```

File: augur/types.py [hunk 3]
```diff
@@ -14,21 +33,22 @@ class DataErrorMethod(enum.Enum):
...
+class EmptyResultsReportingMethod(ArgparseEnum):
     """
-    Enum representation of string values that represent how validation should
-    be handled.
+    Enum representation of string values that represent how empty results should
+    be reported.
     """
     ERROR = 'error'
     WARN  = 'warn'
     SKIP  = 'skip'
```
Comment (huddlej) [inline augur/types.py:43]: This is a minor point, but "skip" doesn't quite match the context in augur filter like it did in the augur validation context. I'd think "ignore" might be a more appropriate verb for the empty filter results.

File: tests/functional/filter/cram/filter-empty-results-reporting.t [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  $ wc -l filtered_strains.txt
+  \s*0 .* (re)
+
+Repeat with the --empty-results-reporting=skip option.
+This should not output any messages to stderr.
+
+  $ ${AUGUR} filter \
+  >  --metadata "$TESTDIR/../data/metadata.tsv" \
+  >  --exclude-all \
+  >  --output-strains filtered_strains.txt \
```
Comment (huddlej) [inline tests/functional/filter/cram/filter-empty-results-reporting.t:36]: Would you mind adding a test (or modifying this one) that outputs empty sequences and metadata along with empty strains? The first behavior I thought to check manually was whether the command also appropriately created empty versions of those outputs, too, but it would be nice if functional tests already covered that.

File: augur/types.py [hunk 5]
```diff
@@ -1,8 +1,27 @@
 import enum
 
 
+class ArgparseEnum(enum.Enum):
+    """
+    Intended to be used as a parent class for any enum representation of
+    string values to be used with argparse options.
+
+    Can be replaced by :py:class:`enum.StrEnum` once Augur's minimum supported
+    Python version is 3.11.
```
Comment (huddlej) [inline augur/types.py:10]: Cool! I learned more about enums from this PR! :D

File: augur/filter/__init__.py [hunk 6]
```diff
@@ -63,11 +64,17 @@ def register_arguments(parser):
...
     output_group.add_argument('--output', '--output-sequences', '-o', help="filtered sequences in FASTA format")
     output_group.add_argument('--output-metadata', help="metadata for strains that passed filters")
     output_group.add_argument('--output-strains', help="list of strains that passed filters (no header)")
     output_group.add_argument('--output-log', help="tab-delimited file with one row for each filtered strain and the reason it was filtered. Keyword arguments used for a given filter are reported in JSON format in a `kwargs` column.")
+    output_group.add_argument(
+        '--empty-results-reporting',
+        type=EmptyResultsReportingMethod,
+        choices=list(EmptyResultsReportingMethod),
+        default=EmptyResultsReportingMethod.ERROR,
+        help="How should empty filter results be reported.")
```
Comment (huddlej) [inline augur/filter/__init__.py:77]: _non-blocking_
Related to my minor comment about using "ignore" instead of "skip", I slightly prefer "handled" to "reported" in the context of "how should empty filter results be...".

File: augur/filter/__init__.py (hunk 1 above)
Comment (victorlin) [reply augur/filter/__init__.py:73]: Looks good!  I like `--empty-results-reporting` over `--empty-output-reporting`.

File: augur/filter/__init__.py [hunk 7]
```diff
@@ -63,11 +64,17 @@ def register_arguments(parser):
...
-    output_group = parser.add_argument_group("outputs", "possible representations of filtered data (at least one required)")
+    output_group = parser.add_argument_group("outputs", "options related to outputs, at least one of the possible representations of filtered data (--output, --output-metadata, --output-strains) is required")
     output_group.add_argument('--output', '--output-sequences', '-o', help="filtered sequences in FASTA format")
     output_group.add_argument('--output-metadata', help="metadata for strains that passed filters")
     output_group.add_argument('--output-strains', help="list of strains that passed filters (no header)")
     output_group.add_argument('--output-log', help="tab-delimited file with one row for each filtered strain and the reason it was filtered. Keyword arguments used for a given filter are reported in JSON format in a `kwargs` column.")
+    output_group.add_argument(
+        '--empty-results-reporting',
+        type=EmptyResultsReportingMethod,
+        choices=list(EmptyResultsReportingMethod),
```
Comment (huddlej) [inline augur/filter/__init__.py:75]: The default argparse behavior when an argument has a list of `choices` and the user provides an invalid choice is to print an error that lists the valid choices. For example, if I run augur filter with `--exclude-ambiguous-dates-by fake`, I get the error:

```
augur filter: error: argument --exclude-ambiguous-dates-by: invalid choice: 'fake' (choose from 'any', 'day', 'month', 'year')
```

I wish that this implementation for results reporting behaved the same way, but instead if I run augur filter with `--empty-results-reporting fake`, I get:

```
augur filter: error: argument --empty-results-reporting: invalid EmptyResultsReportingMethod value: 'fake'
```

I hadn't thought about this specific type of implementation before now, so I checked [the Python argparse docs for `choices`](https://docs.python.org/3/library/argparse.html#choices) for examples and found:

> Use of [enum.Enum](https://docs.python.org/3/library/enum.html#enum.Enum) is not recommended because it is difficult to control its appearance in usage, help, and error messages.

What if we dropped the enum class and used a list of choice strings `["error", "warn", "skip"]` instead? Is there a strong argument to use the enum class instead?

**Review by huddlej — CHANGES_REQUESTED** [review]
Thank you for putting this together, @joverlee521! I confirmed that this new functionality allows the seasonal flu workflow to move past the step where it had previously crashed. In doing so, I realized that we've only kicked the can down the road a bit and now have to make `augur frequencies` handle empty alignment input or modify the workflow to conditionally check for empty inputs. 😅 

My only major request below is for a slightly expanded functional test. Other comments are mostly about naming of things and why we use `enum` instead of a list of choices. I realize that the `enum` pattern predates this PR, but I'm not sure why we'd prefer to use the enum class as an argument type here specifically.

File: augur/filter/__init__.py (hunk 1 above)
Comment (tsibley) [reply augur/filter/__init__.py:73]: _non-blocking_

I think results vs. output is splitting hairs that and no one will be confused by it not applying to `--output-log`. We also don't use "results" anywhere in the `augur filter` help text and nearly nowhere else in Augur's user-facing interface, whereas we do use "output" everywhere for this concept.

File: augur/filter/__init__.py (hunk 6 above)
Comment (tsibley) [reply augur/filter/__init__.py:77]: There's precedent for `--*-reporting` options in `augur curate` commands.

File: augur/types.py (hunk 5 above)
Comment (joverlee521) [reply augur/types.py:10]: 😄

File: tests/functional/filter/cram/filter-empty-results-reporting.t (hunk 4 above)
Comment (joverlee521) [reply tests/functional/filter/cram/filter-empty-results-reporting.t:36]: Thanks for asking for this! Turns out we always write the metadata headers if the metadata output is requested! 

Does including the metadata headers even when results are empty make sense? Or should this behavior be changed to output a completely empty file?

Comment (joverlee521) [reply tests/functional/filter/cram/filter-empty-results-reporting.t:36]: Sounds good! I'll just make sure the header matches the input header.

File: augur/types.py (hunk 3 above)
Comment (huddlej) [reply augur/types.py:43]: I agree that "skip" matches "reporting", but I'm suggesting that "reporting" reflects the internals of the code and not what the user is asking augur filter to do.

As a user, I want to tell augur filter what to do (or how to handle the case) when there are no results in the output: throw an error, print a warning, or do nothing (i.e., ignore the empty output). Asking augur filter to "skip reporting empty results" doesn't quite match how I normally think about responding to errors and warnings (which would be to "ignore empty results"). There might be a better verb than "ignore" or "skip" that captures "do nothing" in a single word though...

By the same logic, asking augur export to "skip validation" matches what I want as a user (when I get the annoying augur version mismatch errors). I don't think I would ask augur export to "ignore validation" because I would want to avoid running validation altogether.

Comment (joverlee521) [reply augur/types.py:43]: Huh, I'm thinking the opposite, where "handled" reflects internals of the code.

As a user, I'm telling augur filter how to report empty results to me. Report it as an error, a warning, or skip the report altogether.

Comment (tsibley) [reply augur/types.py:43]: Ah, well, your examples of what makes sense to you seem inconsistent to me... :upside_down_face: so I think we either have different connotations here or else we're accidentally talking past each other.

This is minor, and I'm happy to drop it in favor of whatever @joverlee521 and you think is best.

File: augur/filter/__init__.py (hunk 7 above)
Comment (joverlee521) [reply augur/filter/__init__.py:75]: The choice of using enums was to make development and maintenance easier for us. 
It helps us standardize the choice values, we no longer have to check arg values with "magic" strings, and I think it makes the code more readable. 

I can update the `ArgparseEnum` class with a custom method to make this error message more useful.

File: augur/types.py (hunk 3 above)
Comment (joverlee521) [reply augur/types.py:43]: > There might be a better verb than "ignore" or "skip" that captures "do nothing" in a single word though...

We can continue the augur curate model and go with `silent` 😄

File: augur/filter/__init__.py (hunk 7 above)
Comment (joverlee521) [reply augur/filter/__init__.py:75]: See custom type converter added in 4decbe46fa1c6d816ff2dca613b811791a2a2737. 

This now prints error as:
```
augur filter: error: argument --empty-results-reporting: invalid choice: 'fake' (choose from 'error', 'warn', 'skip')
```

File: tests/functional/filter/cram/filter-empty-results-reporting.t (hunk 4 above)
Comment (joverlee521) [reply tests/functional/filter/cram/filter-empty-results-reporting.t:40]: Added the additional outputs to the Cram test in force push in 9df8b42fd7037e522f953fe232f5c3b1d8513632.

File: augur/types.py (hunk 3 above)
Comment (huddlej) [reply augur/types.py:60]: Heh, yeah, `silent` could work, too. Maybe it's too minor to dwell on more. I'm happy with whatever you prefer, @joverlee521.

File: augur/filter/__init__.py (hunk 7 above)
Comment (huddlej) [reply augur/filter/__init__.py:75]: Got it! Thank you for the improved error message, @joverlee521. It looks like it wasn't quite as bad as the Python docs made it sound, after all!

File: augur/filter/__init__.py (hunk 1 above)
Comment (corneliusroemer) [reply augur/filter/__init__.py:73]: I like the resolution, the rewording of 

```
possible representations of filtered data (at least one required)
```
to 

```
options related to outputs, at least one of the possible representations
  of filtered data (--output, --output-metadata, --output-strains) is
  required`
```

is great!

**Review by huddlej — APPROVED** [review]

File: augur/filter/__init__.py (hunk 1 above)
Comment (victorlin) [reply augur/filter/__init__.py:73]: > We also don't use "results" anywhere in the `augur filter` help text and nearly nowhere else in Augur's user-facing interface, whereas we do use "output" everywhere for this concept.

This is a good point. I'm on board with `--empty-output-reporting` now.

File: augur/types.py [hunk 8]
```diff
@@ -19,6 +20,22 @@ def __str__(self) -> str:
...
+        Intended to be used as the argument type converter for argparse options
+        that use the enum values as inputs.
+
+        Raises a custom `argparse.ArgumentTypeError` so that the error
+        message can include a helpful list of the all valid enum values.
+        """
+        try:
+            return cls(input_string)
+        except ValueError as error:
+            choices = ", ".join(map(lambda x: f"'{str(x)}'", cls))
```
Comment (tsibley) [inline augur/types.py:35]: _non-blocking_

```suggestion
            choices = ", ".join(f"{str(x)!r}" for x in cls)
```

Or similar? Main point here is to use `repr()` (here via `!r`) instead of manually surrounding with single quotes.

File: augur/filter/__init__.py (hunk 1 above)
Comment (joverlee521) [reply augur/filter/__init__.py]: Thanks for the all the feedback! I updated the option to `--empty-output-reporting` and reworded the help message to be clear that the option applies to the filtered data:
```
  --empty-output-reporting {error,warn,silent}
                        How should empty outputs be reported when no strains
                        pass filtering and/or subsampling. (default: error)
```

File: augur/types.py (hunk 3 above)
Comment (joverlee521) [reply augur/types.py]: Updated to `silent` in the forced-push.

File: augur/types.py (hunk 8 above)
Comment (joverlee521) [reply augur/types.py]: Thanks! Updated in the forced-push.
